File: plugins/gumshoe/base/ui/src/types/obstacle.ts
import { SkillsConfig } from "./skills";

export type ObstacleClue = {
  type: 'clue';
  name: string;
  description?: string;

  // список investigative skills, которыми можно получить/купить улику
  investigative_skills: string[];
  spend_cost: number; // >= 0; 0 = core clue / бесплатно
  reward?: string;
};

export type ObstacleChallenge = {
  type: 'challenge';
  name: string;
  description?: string;

  general_skill: string; // id general skill
  difficulty: number;    // >= 2 (на бэке PositiveInt, но можно ограничить в UI)
  on_success?: string;
  on_fail?: string;
};

export type ObstacleData = ObstacleClue | ObstacleChallenge;


export type ObstacleConfig = SkillsConfig & {
  initialData: ObstacleData;
};

File: plugins/gumshoe/base/backend/types/obstacle.py
# types/obstacle.py
from __future__ import annotations

from typing import Annotated, Literal, Union
from pydantic import BaseModel, Field, NonNegativeInt, PositiveInt


class ObstacleClue(BaseModel):
    type: Literal["clue"] = "clue"
    investigative_skills: list[str] = Field(default_factory=list)
    spend_cost: NonNegativeInt = 0
    reward: str = ""


class ObstacleChallenge(BaseModel):
    type: Literal["challenge"] = "challenge"
    general_skill: str = ""
    difficulty: PositiveInt = 4
    on_success: str = ""
    on_fail: str = ""


ObstacleData = Annotated[
    Union[ObstacleClue, ObstacleChallenge],
    Field(discriminator="type"),
]
